cpp/src/arrow/io/mremap.h

```diff
@@ -0,0 +1,80 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+#ifdef _WIN32
+#include "arrow/io/mman.h"
+#undef Realloc
+#undef Free
+#else
+#include <sys/mman.h>
+#include <unistd.h>
+#endif
+
+// define a wrapper around mremap which the same arguments on linux and on windows
+// should only be called with writable files
+static void* arrow_mremap(void* addr, size_t old_size, size_t new_size, int fildes) {
+#ifdef _WIN32
+  // flags are ignored on windows
+  void* new_addr = MAP_FAILED;
+  HANDLE fm, h;
+  DWORD dwErrCode = 0;
+
+  /* First, unmap the file view */
```

Comment:
This comment convention is different from the others. That made me think there might be some copy-pasting here from another codebase

https://github.com/search?q=%22+First%2C+unmap+the+file+view%22&type=Code

If so, which codebase did the code come from? We need to properly document the IP. Let us know if you reused code from anywhere else in this patch